server-spi-private/src/main/java/org/keycloak/compatibility/CompatibilityMetadataProvider.java

```diff
@@ -0,0 +1,51 @@
+package org.keycloak.compatibility;
+
+import java.util.Map;
+
+/**
+ * Provides the metadata used by the "update-compatibility" command.
+ * <p>
+ * Implementations should return all metadata required to determine if it is possible to update from one Keycloak
+ * deployment to another in a compatible manner. Metadata key/value pairs may be added or removed in a subsequent version,
+ * so it's necessary for implementations to handle missing metadata gracefully.
+ * <p>
+ * The {@link CompatibilityResult} determines if a rolling update is possible. Factory methods are present with default
+ * implementations of {@link CompatibilityResult}.
+ */
+public interface CompatibilityMetadataProvider {
+
+    int DEFAULT_PRIORITY = 1;
+
+    /**
+     * @return The metadata required by this provider to determine if a rolling update is possible.
+     */
+    Map<String, String> metadata();
```

Comment:
Are we sure `Map<String, String>` is enough? Do we want to consider cases when some provider would like to have some nested object inside value? We could have `Map<String, Object>` instead which would allow providers to choose any structure they want for their metadata. 

One real life example that this could use: Imagine we would want to have configuraiton for all SPIs in this file:
`SPI_name` -> `SPI_provider_id` -> `config option name` -> `config option value`

Sure we could have first three things inside key but the file would be more readable with nested objects. 

WDYT?

Reply:
Yes, it is enough.

I considered `Map<String, Object>`, but Jackson needs to know the class type for deserialization. The interface would have to be typed, and the class would have to be provided to Keycloak. It complicates everything with zero benefit.

The implementation can use any kind of object they want internally. They just have to use Jackson to convert it to a single string or map of strings. Appending all the components in the map's key is also a solution and probably simpler.

My conclusion, keep things simpler :+1:

Reply:
Ok, agreed.